File: python/common/src/recoveryai_common/observability/tracing.py
"""Distributed tracing for Python services (plan Section 4.3).

W3C `traceparent` is the only propagator, so a trace started in a TypeScript
service continues here and back without translation.
"""

from collections.abc import AsyncIterator
from contextlib import asynccontextmanager

import httpx
from opentelemetry import propagate, trace
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator

_provider: TracerProvider | None = None


def init_tracing(
    *,
    service: str,
    environment: str,
    endpoint: str | None,
    enabled: bool = True,
) -> None:
    """Install the tracer provider. Disabling tracing must never break the service."""
    global _provider
    if _provider is not None:
        return

    propagate.set_global_textmap(TraceContextTextMapPropagator())
    if not enabled:
        return

    _provider = TracerProvider(
        resource=Resource.create(
            {
                "service.name": service,
                "service.version": "0.0.0",
                "deployment.environment.name": environment,
            }
        )
    )
    if endpoint:
        _provider.add_span_processor(
            BatchSpanProcessor(OTLPSpanExporter(endpoint=f"{endpoint.rstrip('/')}/v1/traces"))
        )
    trace.set_tracer_provider(_provider)


def shutdown_tracing() -> None:
    global _provider
    if _provider is None:
        return
    _provider.shutdown()
    _provider = None


def get_tracer(name: str) -> trace.Tracer:
    return trace.get_tracer(name)


@asynccontextmanager
async def traced_client(**kwargs: object) -> AsyncIterator[httpx.AsyncClient]:
    """An httpx client that injects the active trace context into outgoing requests."""

    async def inject(request: httpx.Request) -> None:
        carrier: dict[str, str] = {}
        propagate.inject(carrier)
        for key, value in carrier.items():
            request.headers[key] = value

    async with httpx.AsyncClient(event_hooks={"request": [inject]}, **kwargs) as client:  # type: ignore[arg-type]
        yield client

File: packages/observability-ts/src/tracing.ts
import { type Tracer, context, propagation, trace } from '@opentelemetry/api';
import { W3CTraceContextPropagator } from '@opentelemetry/core';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http';
import { resourceFromAttributes } from '@opentelemetry/resources';
import { BatchSpanProcessor, type SpanProcessor } from '@opentelemetry/sdk-trace-base';
import { NodeTracerProvider } from '@opentelemetry/sdk-trace-node';
import { ATTR_SERVICE_NAME, ATTR_SERVICE_VERSION } from '@opentelemetry/semantic-conventions';

let provider: NodeTracerProvider | undefined;

export interface TracingOptions {
  service: string;
  environment: string;
  endpoint?: string | undefined;
  enabled: boolean;
}

/**
 * Distributed tracing bootstrap (plan Section 4.3).
 *
 * W3C `traceparent` is the only propagator, so a trace started in a TypeScript
 * service continues into the Python services and back without translation.
 * Disabling tracing is supported and must never break the service.
 */
export function initTracing(options: TracingOptions): void {
  if (provider) return;
  if (!options.enabled) {
    propagation.setGlobalPropagator(new W3CTraceContextPropagator());
    return;
  }

  const spanProcessors: SpanProcessor[] = [];
  if (options.endpoint) {
    spanProcessors.push(
      new BatchSpanProcessor(
        new OTLPTraceExporter({ url: `${options.endpoint.replace(/\/$/, '')}/v1/traces` }),
      ),
    );
  }

  provider = new NodeTracerProvider({
    resource: resourceFromAttributes({
      [ATTR_SERVICE_NAME]: options.service,
      [ATTR_SERVICE_VERSION]: '0.0.0',
      'deployment.environment.name': options.environment,
    }),
    spanProcessors,
  });

  provider.register({ propagator: new W3CTraceContextPropagator() });
}

export async function shutdownTracing(): Promise<void> {
  if (!provider) return;
  await provider.shutdown();
  provider = undefined;
}

export function getTracer(name: string): Tracer {
  return trace.getTracer(name);
}

/**
 * `fetch` that injects the active trace context into outgoing headers, so an
 * internal service-to-service call joins the caller's trace (plan Section 4.3).
 */
export async function tracedFetch(input: string | URL | Request, init?: RequestInit) {
  const headers = new Headers(init?.headers ?? {});
  const carrier: Record<string, string> = {};
  propagation.inject(context.active(), carrier);
  for (const [key, value] of Object.entries(carrier)) headers.set(key, value);
  return fetch(input, { ...init, headers });
}
